🌱 Improve dry run for topology changes to dry run server side apply (#6710)
kubernetes-sigs/cluster-api: docs/book/src/developer/providers/v1.1-to-v1.2.md

  [merge-strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
  NOTE: the change will cause a rollout on existing clusters created with ClusterClass

  E.g. in CAPA

  ```go
  // +optional
  Subnets Subnets `json:"subnets,omitempty"
  ```
  Must be modified into:

  ```go
  // +optional
  // +listType=map
  // +listMapKey=id
  Subnets Subnets `json:"subnets,omitempty"
  ```

- [Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) implementation in ClusterClass and managed topologies 

>   requires to dry-run changes on templates, and it is required to implement following changes in order to prevent dry-run
>   to return errors due to immutability checks implemented in the web hooks.

Reviewer: ```suggestion
  requires to dry-run changes on templates. If infrastructure or bootstrap providers have implemented immutability checks in their 
  InfrastructureMachineTemplate or BootstrapConfigTemplate webhooks, it is required to implement the following changes in order to prevent dry-run
  to return errors.
```

I think we have to add when it is required. Only a very small subset of infra providers have implemented the immutability check and it isn't documented / standardized anywhere as far as I can tell.

Q: Should we have a follow-up to document this somewhere in the contract? I think we can't expect folks to read through previous migration docs.

Author: I've adjusted the sentence but will leave this unresolved due to docs.

I'm +1 adding thsi to the contract.

Reviewer: I think we can link to there but in the contract doc it's probably framed slightly different (not as a diff, but rather as what to do in ValidateUpdate if you want to do immutability checks)

Author: Maybe add a sentence like this:

```

Validating webhooks implemented for InfrastructureMachineTemplate or BootstrapConfigTemplate are required to not block due to immutability checks when the controller for managed topology and ClusterClass does
[Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) dry-run requests.
```

to https://main.cluster-api.sigs.k8s.io/developer/providers/webhooks.html#validating-webhooks ? wdyt?

Reviewer: That definitely doesn't hurt. I wonder if we should add a code snippet /example for ValidateUpdate + if it should be here instead: https://main.cluster-api.sigs.k8s.io/developer/providers/contracts.html

Reviewer: With some restriction that it's only required if ClusterClass should be supported (similar to what we have here for using versions, using machines, ..): https://main.cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html#required-spec-fields-for-implementations-using-version

Probably makes sense to hear Fabrizio's take on it

  +     newObj, ok := newRaw.(*DockerMachineTemplate)
  +     if !ok {
  +         return apierrors.NewBadRequest(fmt.Sprintf("expected a DockerMachineTemplate but got a %T", newRaw))
  +     }
  - func (m *DockerMachineTemplate) ValidateUpdate(oldRaw runtime.Object) error {
        oldObj, ok := oldRaw.(*DockerMachineTemplate)
        if !ok {
            return apierrors.NewBadRequest(fmt.Sprintf("expected a DockerMachineTemplate but got a %T", oldRaw))
        }
        ...
        // Immutability check
  +     if !topology.ShouldSkipImmutabilityChecks(ctx, newObj) &&
  +        !reflect.DeepEqual(newObj.Spec.Template.Spec, oldObj.Spec.Template.Spec) {
  -     if !reflect.DeepEqual(m.Spec.Template.Spec, old.Spec.Template.Spec) {
            allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "template", "spec"), m, dockerMachineTemplateImmutableMsg))
        }
        ...
    }

  + func (*DockerMachineTemplateWebhook) ValidateDelete(ctx context.Context, _ runtime.Object) error {
  - func (m *DockerMachineTemplate) ValidateDelete() error {
        ...
    }
  ```

NOTES:

> - We are introducing a `DockerMachineTemplateWebhook` struct because we are going to use a controller runtime
>   `CustomValidator` that will allow to skip the immutability check only when the topology controller is dry running
>   while preserving the validation behaviour for all the other cases.

Reviewer: ```suggestion
- We are introducing a `DockerMachineTemplateWebhook` struct because we are going to use a controller runtime
  `CustomValidator`. This will allow to skip the immutability check only when the topology controller is dry running
  while preserving the validation behaviour for all other cases.
```
nit

> - By using CustomValidators it will be possible to move webhooks to other packages, thus removing controller
>   runtime dependency from the API types. However, choosing to do so or not is up to the providers implementers. 

Reviewer: ```suggestion
- By using `CustomValidators` it is possible to move webhooks to other packages, thus removing some controller
  runtime dependency from the API types. However, choosing to do so or not is up to the provider implementers
  and independent of this change. 
```
nit

  [merge-strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy).
  NOTE: the change will cause a rollout on existing clusters created with ClusterClass

  E.g. in CAPA

  ```go
  // +optional
  Subnets Subnets `json:"subnets,omitempty"
  ```
  Must be modified into:

  ```go
  // +optional
  // +listType=map
  // +listMapKey=id
  Subnets Subnets `json:"subnets,omitempty"
  ```

> - [Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) implementation in ClusterClass and managed topologies 

Reviewer: Should we add an info that controller-runtime >= v0.12.3 is required?

Reviewer 2: It could be useful to add controller runtime to the list on top of the page as well
